File: php/test/MessageEntityTest.php
<?php
declare(strict_types=1);

// Message entity test

require_once __DIR__ . '/../thesmsworks_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class MessageEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = ThesmsworksSDK::test(null, null);
        $ent = $testsdk->Message(null);
        $this->assertNotNull($ent);
    }

    public function test_basic_flow(): void
    {
        $setup = message_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["create", "load", "remove"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "message." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set THESMSWORKS_TEST_MESSAGE_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // CREATE
        $message_ref01_ent = $client->Message(null);
        $message_ref01_data = Helpers::to_map(Vs::getprop(
            Vs::getpath($setup["data"], "new.message"), "message_ref01"));

        $message_ref01_data_result = $message_ref01_ent->create($message_ref01_data, null);
        $message_ref01_data = Helpers::to_map(is_object($message_ref01_data_result) && method_exists($message_ref01_data_result, 'data_get') ? $message_ref01_data_result->data_get() : $message_ref01_data_result);
        $this->assertNotNull($message_ref01_data);
        $this->assertNotNull($message_ref01_data["id"]);

        // LOAD
        $message_ref01_match_dt0 = [
            "id" => $message_ref01_data["id"],
        ];
        $message_ref01_data_dt0_loaded = $message_ref01_ent->load($message_ref01_match_dt0, null);
        $message_ref01_data_dt0_load_result = Helpers::to_map(is_object($message_ref01_data_dt0_loaded) && method_exists($message_ref01_data_dt0_loaded, 'data_get') ? $message_ref01_data_dt0_loaded->data_get() : $message_ref01_data_dt0_loaded);
        $this->assertNotNull($message_ref01_data_dt0_load_result);
        $this->assertEquals($message_ref01_data_dt0_load_result["id"], $message_ref01_data["id"]);

        // REMOVE
        $message_ref01_match_rm0 = [
            "id" => $message_ref01_data["id"],
        ];
        $message_ref01_ent->remove($message_ref01_match_rm0, null);

    }
}

function message_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/message/MessageTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = ThesmsworksSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["message01", "message02", "message03", "schedule01", "schedule02", "schedule03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("THESMSWORKS_TEST_MESSAGE_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "THESMSWORKS_TEST_MESSAGE_ENTID" => $idmap,
        "THESMSWORKS_TEST_LIVE" => "FALSE",
        "THESMSWORKS_TEST_EXPLAIN" => "FALSE",
        "THESMSWORKS_APIKEY" => "",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["THESMSWORKS_TEST_MESSAGE_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["THESMSWORKS_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            // FIRST, so the generated fields below win: sdk-test-control.json's
            // test.client.options adds to the live client, it does not redirect it.
            Runner::live_client_options(),
            [
                "apikey" => $env["THESMSWORKS_APIKEY"],
            ],
            // ismap, not a plain "?? []" default: an empty PHP array is a
            // LIST, and a non-map later entry REPLACES the accumulated map in
            // merge - so the no-extras call discarded live_client_options()
            // and the apikey/server map above it.
            Vs::ismap($extra) ? $extra : new \stdClass(),
        ]);
        // "?? []" because merge legitimately answers with a stdClass when every
        // contributing entry is an EMPTY map - an SDK with no apikey and no
        // server variables generates an empty middle entry, so that is the
        // common case, not the edge one. to_map returns null for a non-array by
        // design, and the constructor takes a non-nullable array, so without the
        // fallback every such SDK died on "must be of type array, null given"
        // the moment live mode was switched on. Offline mode never reaches this
        // branch, which is why the offline suite stayed green.
        $client = new ThesmsworksSDK(Helpers::to_map($merged_opts) ?? []);
    }

    $live = $env["THESMSWORKS_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["THESMSWORKS_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}

File: kotlin/test/MessageEntityTest.kt
package voxgig.thesmsworkssdk.sdktest

import java.nio.file.Files
import java.nio.file.Paths

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Assumptions
import org.junit.jupiter.api.Test

import voxgig.thesmsworkssdk.core.Helpers
import voxgig.thesmsworkssdk.core.SdkEntity
import voxgig.thesmsworkssdk.core.ThesmsworksSDK
import voxgig.thesmsworkssdk.utility.Json
import voxgig.thesmsworkssdk.utility.struct.Struct

@Suppress("UNCHECKED_CAST", "UNUSED_VARIABLE", "UNUSED_VALUE")
class MessageEntityTest {

  @Test
  fun instance() {
    val testsdk = ThesmsworksSDK.testSDK()
    val ent = testsdk.message(null)
    assertNotNull(ent, "expected non-null message entity")
  }

  @Test
  fun basic() {
    val setup = messageBasicSetup(null)
    // Per-op sdk-test-control.json skip.
    val mode = if (setup.live) "live" else "unit"
    for (op in arrayOf("create", "load", "remove")) {
      val reason = RunnerSupport.skipReason("entityOp", "message.$op", mode)
      Assumptions.assumeTrue(
        reason == null,
        if (reason == null || "" == reason) "skipped via sdk-test-control.json" else reason,
      )
    }
    Assumptions.assumeFalse(
      setup.syntheticOnly,
      "live entity test uses synthetic IDs from fixture — set THESMSWORKS_TEST_MESSAGE_ENTID JSON to run live",
    )
    val client = setup.client

    // CREATE
    val messageRef01Ent = client.message(null)
    var messageRef01Data: MutableMap<String, Any?> = (Helpers.toMapAny(Struct.getprop(
        Struct.getpath(setup.data, "new.message"), "message_ref01")) ?: linkedMapOf())

    val messageRef01DataResult = messageRef01Ent.create(messageRef01Data, null)
    messageRef01Data = Helpers.toMapAny(messageRef01DataResult) ?: linkedMapOf()
    assertNotNull(messageRef01Data, "expected create result to be a map")

    // LOAD
    val messageRef01MatchDt0 = linkedMapOf<String, Any?>()
    val messageRef01DataDt0Loaded = messageRef01Ent.load(messageRef01MatchDt0, null)
    assertNotNull(messageRef01DataDt0Loaded, "expected load result to be non-null")

    // REMOVE
    val messageRef01MatchRm0 = linkedMapOf<String, Any?>()
    messageRef01MatchRm0["id"] = messageRef01Data["id"]
    messageRef01Ent.remove(messageRef01MatchRm0, null)

  }

  companion object {
    fun messageBasicSetup(extra: MutableMap<String, Any?>?): RunnerSupport.EntityTestSetup {
      RunnerSupport.loadEnvLocal()

      val entityData: MutableMap<String, Any?>
      try {
        val entityDataSource = Files.readString(Paths.get(
            "..", ".sdk", "test", "entity", "message", "MessageTestData.json"))
        entityData = Helpers.toMapAny(Json.parse(entityDataSource)) ?: linkedMapOf()
      } catch (e: Exception) {
        throw AssertionError("failed to read message test data: " + e.message, e)
      }

      val options = linkedMapOf<String, Any?>()
      options["entity"] = entityData["existing"]

      var client = ThesmsworksSDK.testSDK(options, extra)

      // Generate idmap via transform, matching TS pattern.
      val idnames = mutableListOf<Any?>()
      idnames.add("message01")
      idnames.add("message02")
      idnames.add("message03")
      idnames.add("schedule01")
      idnames.add("schedule02")
      idnames.add("schedule03")
      val idmap = Struct.transform(idnames, Json.parse(
          "{\"`\$PACK`\": [\"\", {" +
          "\"`\$KEY`\": \"`\$COPY`\"," +
          "\"`\$VAL`\": [\"`\$FORMAT`\", \"upper\", \"`\$COPY`\"]" +
          "}]}"))

      // Detect ENTID env override before envOverride consumes it.
      val entidEnvRaw = RunnerSupport.getenv("THESMSWORKS_TEST_MESSAGE_ENTID")
      val idmapOverridden = entidEnvRaw != null && entidEnvRaw.trim().startsWith("{")

      val envm = linkedMapOf<String, Any?>()
      envm["THESMSWORKS_TEST_MESSAGE_ENTID"] = idmap
      envm["THESMSWORKS_TEST_LIVE"] = "FALSE"
      envm["THESMSWORKS_TEST_EXPLAIN"] = "FALSE"
      envm["THESMSWORKS_APIKEY"] = "NONE"
      val env = RunnerSupport.envOverride(envm)

      var idmapResolved = Helpers.toMapAny(env["THESMSWORKS_TEST_MESSAGE_ENTID"])
      if (idmapResolved == null) {
        idmapResolved = Helpers.toMapAny(idmap) ?: linkedMapOf()
      }

      val live = "TRUE" == env["THESMSWORKS_TEST_LIVE"]
      if (live) {
        val liveOpts = linkedMapOf<String, Any?>()
        liveOpts["apikey"] = env["THESMSWORKS_APIKEY"]
        val mergedOpts = Struct.merge(Struct.jt(liveOpts, extra))
        client = ThesmsworksSDK(Helpers.toMapAny(mergedOpts))
      }

      val setup = RunnerSupport.EntityTestSetup()
      setup.client = client
      setup.data = entityData
      setup.idmap = idmapResolved
      setup.env = env
      setup.explain = "TRUE" == env["THESMSWORKS_TEST_EXPLAIN"]
      setup.live = live
      setup.syntheticOnly = live && !idmapOverridden
      setup.now = System.currentTimeMillis()
      return setup
    }
  }
}
